pgx/src/pgx.ml

```diff
@@ -526,7 +531,59 @@ module Make (Thread : Io) = struct
 
   (*----- Connection. -----*)
 
+  let attempt_tls_upgrade ?(ssl = `Auto) ({ ichan; chan; _ } as conn) =
+    (* To initiate an SSL-encrypted connection, the frontend initially sends an SSLRequest message rather than a
+       StartupMessage. The server then responds with a single byte containing S or N, indicating that it is willing
+       or unwilling to perform SSL, respectively. The frontend might close the connection at this point if it is
+        dissatisfied with the response. To continue after S, perform an SSL startup handshake (not described here,
+       part of the SSL specification) with the server. If this is successful, continue with sending the usual
+       StartupMessage. In this case the StartupMessage and all subsequent data will be SSL-encrypted. To continue
+       after N, send the usual StartupMessage and proceed without encryption.
+       See https://www.postgresql.org/docs/9.3/protocol-flow.html#AEN100021 *)
+    match ssl with
+    | `No -> return conn
+    | (`Auto | `Always _) as ssl ->
+      (match Io.upgrade_ssl with
+      | `Not_supported ->
+        (match ssl with
+        | `Always _ ->
+          failwith
+            "TLS support is not compiled into this Pgx library but ~ssl was set to \
+             `Always"
```

Comment:
I'd like to make this impossible but I'm not sure how.

Reply:
#suggestion: We can also enforce that the io wrappers for pgx should always pick an ssl option (async_ssl, tls or lwt_ssl), and get rid of this branch. I'd expect that eventually all io wrappers in the pgx repo will support encrypted connections, but if we don't want to support them in a particular backend, we can probably tweak the wrapper itself to not expose any ssl related options in its mli file